File: C++/단체사진.cpp
#include <string>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int max_count;

void loop(int n, int clear, int count, vector<string> data, vector<string> kakao_friends, vector<string> line) {

    line.push_back(kakao_friends[count]);   // 새로운 벡터에 추가
    kakao_friends.erase(kakao_friends.begin() + count);     // 친구 벡터에서 제거

    if (kakao_friends.size() == 0) {
        int check = 0;  // 두명의 친구가 line에 존재하는지
        for (int i = 0; i < n; i++) {
            string cmp1, cmp2;  //algorithm의 find 함수를 사용할때 line의 string과 형식을 같이 하기 위해 만든 변수
                cmp1 = data[i][0];  // char to string
                cmp2 = data[i][2];  // char to string
                int item_cmp = abs((find(line.begin(), line.end(), cmp1) - line.begin()) - (find(line.begin(), line.end(), cmp2) - line.begin()));  // 한조건의 두친구 간격구하기
                int distance = data[i][4] - 47; // 두친구의 조건 거리
                if (data[i][3] == '<') {           // 두친구의 조건
                    if (item_cmp < distance) {
                        check++;
                    }
                }
                else if (data[i][3] == '>') {
                    if (item_cmp > distance) {
                        check++;
                    }

                }
                else if (data[i][3] == '=') {
                    if (item_cmp == distance) {
                        check++;
                    }
                }
            }
            if (check == 2) {   // 두조건에 만족했다면
                max_count++;
            }
        
    }
    for (int i = 0; i < kakao_friends.size(); i++) {        // 한무 함수
        loop(n, clear, i, data, kakao_friends, line);
    }
}


// 이코드는 글렀나봄.... 다시 만들어보자

// 전역 변수를 정의할 경우 함수 내에 초기화 코드를 꼭 작성해주세요.
int solution(int n, vector<string> data) {     
    int answer = 0;
    vector<string> kakao_friends= { "A","C","F", "J", "M", "N", "R","T" };
    vector<string> line;
    for (int i = 0; i < kakao_friends.size(); i++) {
        loop(n, 0, i, data, kakao_friends, line);
    }
    answer = max_count;
    return answer;
}

File: Java/단체사진.java
class Solution {
    String[] kakao_friends = {"A", "C", "F", "J", "M", "N", "R", "T"};      // 카카오 프렌즈
    int answer = 0;     // 결과

    public int solution(int n, String[] data) {
        boolean[] visited = new boolean[8];     // 방분 여부 배열
        search("", visited, n, data);           // 검색 재귀함수
        return answer;
    }

    public void search( String str, boolean[] visited, int n, String[] data){
        if(str.length() == 8){      // 잘 이해가 안가지만 7을 넣어도 답은 나오고 오히려 횟수가 줄어 성능이 좋다 : 마지막 길이까지 도달 했을 경우
            boolean check = true;   // 조건이 맞는지 확인하는 함수
            int ac,distance;        // ac : 계산값, distance : 두 프렌즈간 거리
            char option;            // 거리 조건
            for(String line : data) {   // 조건 만큼 반복
                ac = Math.abs(str.indexOf(line.charAt(0)) - str.indexOf(line.charAt(2)));   // 절대값(두 프렌즈간 거리)
                option = line.charAt(3);            // 조건 = < >
                distance = line.charAt(4) - 47;     // 프렌즈간 거리 ex) 1번째 0, 2번째 2
                if (option == '=') {        //조건 =
                    if (!(ac == distance)) {//조건과 상이 하면 false, 그 뒤는 무의미 하기 때문에 break
                        check = false;
                        break;
                    }
                } else if (option == '<') { //조건 < 
                    if (!(ac < distance)) { 
                        check = false;
                        break;
                    }
                } else if (option == '>') { //조건 < 
                    if (!(ac > distance)) { 
                        check = false;
                        break;
                    }
                }
            }
            if(check){      // 모든 조건을 통과하면 비로소 카운트
                answer++;
            }
            return;
        }
        for(int num = 0; num < kakao_friends.length ; num++){   //프렌즈의 수만큼 반복
            if(!visited[num]){          // 방분하지 않았다면
                visited[num] = true;    // 방문했다고 알리고 재귀함수 진입
                search( str + kakao_friends[num], visited, n, data);
                visited[num] = false;   // 재귀함수가 끝나면 나간다고 알림
            }
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        int n = 2;
        String[] data = {"N~F=0", "R~T>2"};
        System.out.println(s.solution(n,data));
    }
}
